File: apps/ios/Packages/DropCore/Sources/DropCore/NoteHierarchy.swift
import Foundation

/// 목록 화면의 한 줄. 노트 하나와 그 노트를 어떻게 그릴지가 함께 담긴다.
///
/// 답글을 그리려면 노트만으로는 부족하다 — 몇 단 들여쓸지, 맥락으로 끌어온 것인지,
/// 부모를 잃은 것인지를 화면이 알아야 한다. 그 판단을 화면에 두면 검증할 수 없어
/// 여기서 미리 정해 내려보낸다.
public struct NoteRow: Sendable, Equatable, Identifiable {
    public let note: Note
    /// 들여쓰기 단수. 데이터상 깊이가 아니라 **그릴 깊이**다 (상한에서 멈춘다).
    public let depth: Int
    /// 필터·검색에 걸린 것이 아니라 자식의 맥락을 위해 끌어온 노트.
    public let isContextOnly: Bool
    /// 부모가 이 뷰에 없어 최상위로 올라온 답글. 독립 노트처럼 보이면 안 된다.
    public let isOrphanedReply: Bool

    public var id: String { note.id }

    public init(note: Note, depth: Int, isContextOnly: Bool = false, isOrphanedReply: Bool = false) {
        self.note = note
        self.depth = depth
        self.isContextOnly = isContextOnly
        self.isOrphanedReply = isOrphanedReply
    }
}

/// 평평한 노트 목록을 부모-자식으로 묶어 화면이 그대로 그릴 행 배열로 만든다.
///
/// 순수 함수로 떼어 두어 시뮬레이터 없이 검증한다 — 데스크톱은 이 로직이 화면
/// 컴포넌트(`NoteFeed.tsx`) 안에 있어 테스트가 없다. 같은 실수를 반복하지 않는다.
public enum NoteHierarchy {
    /// 화면이 그릴 기본 들여쓰기 상한. 좁은 화면에서 3단 이상 들여쓰면 본문 폭이 무너진다.
    public static let defaultMaxIndentDepth = 2

    /// - Parameters:
    ///   - visible: 필터·검색까지 통과해 **실제로 보여야 하는** 노트. 순서가 최상위 노트의 순서가 된다.
    ///   - context: 같은 뷰(활성/보관/휴지통)에 있는 노트 전부. 부모를 끌어올 후보다.
    ///     `visible`은 이 집합의 부분집합이어야 한다.
    ///   - maxIndentDepth: 들여쓰기 상한.
    ///
    /// 규칙:
    /// 1. 보이는 답글의 부모는 필터에 걸리지 않아도 **끌어온다** — 그러지 않으면 답글이
    ///    최상위로 튀어 어느 노트에 달린 것인지 알 수 없어진다.
    /// 2. 부모가 이 뷰에 아예 없으면(보관·휴지통에 있거나 지워졌으면) 답글을 버리지 않고
    ///    최상위로 올리되 `isOrphanedReply`로 표시한다. 버리면 노트가 조용히 사라진다.
    /// 3. 형제 답글은 오래된 것부터 — 데스크톱(`NoteFeed.tsx`)과 같은 규칙이다.
    public static func rows(
        visible: [Note],
        context: [Note],
        maxIndentDepth: Int = defaultMaxIndentDepth
    ) -> [NoteRow] {
        guard !visible.isEmpty else { return [] }

        // 같은 노트가 두 번 들어와도 행은 한 번만 나야 한다.
        var contextByID: [String: Note] = [:]
        var contextOrder: [String] = []
        for note in context where contextByID[note.id] == nil {
            contextByID[note.id] = note
            contextOrder.append(note.id)
        }

        var visibleIDs: Set<String> = []
        var visibleOrder: [String] = []
        for note in visible where !visibleIDs.contains(note.id) {
            visibleIDs.insert(note.id)
            visibleOrder.append(note.id)
            // visible이 context에 없을 수도 있다(호출자가 다른 집합을 넘긴 경우).
            // 그래도 노트를 잃지 않도록 여기서 채워 넣는다.
            if contextByID[note.id] == nil {
                contextByID[note.id] = note
                contextOrder.append(note.id)
            }
        }

        // 규칙 1 — 보이는 노트의 조상을 맥락으로 끌어온다.
        var included = visibleIDs
        for id in visibleOrder {
            var seen: Set<String> = [id]
            var current = contextByID[id]
            while let parentID = current?.parentID, let parent = contextByID[parentID] {
                // 망가진 데이터(자기 참조·순환)에서 멈추지 않으면 무한 루프가 된다.
                guard seen.insert(parent.id).inserted else { break }
                included.insert(parent.id)
                current = parent
            }
        }

        // 부모를 따라 올라가다 제자리로 돌아오는 노트(망가진 데이터)를 가려낸다.
        // 이것을 자식으로 붙이면 어느 뿌리에서도 닿지 못해 목록에서 조용히 사라진다.
        var loops: Set<String> = []
        for id in contextOrder where included.contains(id) {
            var seen: Set<String> = [id]
            var current = contextByID[id]
            while let parentID = current?.parentID, included.contains(parentID), let parent = contextByID[parentID] {
                guard seen.insert(parent.id).inserted else {
                    loops.insert(id)
                    break
                }
                current = parent
            }
        }

        var childrenByParentID: [String: [Note]] = [:]
        var rootIDs: [String] = []
        for id in contextOrder where included.contains(id) {
            guard let note = contextByID[id] else { continue }
            if let parentID = note.parentID, !loops.contains(id), included.contains(parentID) {
                childrenByParentID[parentID, default: []].append(note)
            } else {
                rootIDs.append(id)
            }
        }

        // 규칙 3 — 형제는 오래된 것부터.
        for (parentID, children) in childrenByParentID {
            childrenByParentID[parentID] = children.sorted { $0.createdAt < $1.createdAt }
        }

        var emitted: Set<String> = []
        var result: [NoteRow] = []

        func emit(_ note: Note, depth: Int) {
            guard emitted.insert(note.id).inserted else { return }
            result.append(
                NoteRow(
                    note: note,
                    depth: min(depth, maxIndentDepth),
                    isContextOnly: !visibleIDs.contains(note.id),
                    // 규칙 2 — 부모가 있는데 최상위로 나온 것은 부모를 잃은 답글이다.
                    isOrphanedReply: depth == 0 && note.parentID != nil && note.parentID != note.id
                )
            )
            for child in childrenByParentID[note.id] ?? [] {
                emit(child, depth: depth + 1)
            }
        }

        for id in rootIDs {
            guard let note = contextByID[id] else { continue }
            emit(note, depth: 0)
        }

        return result
    }
}

File: apps/desktop/src/renderer/lib/note-hierarchy.ts
// 노트 목록을 부모-자식으로 묶어 피드가 그대로 그릴 행 배열로 만든다 (BRU-70).
//
// 원래 이 로직은 NoteFeed의 useMemo 안에 있었다. 화면 안에 있으면 테스트할 방법이
// 없어서, 부모가 필터에서 빠지면 답글이 통째로 사라지는 버그를 아무도 못 잡았다.
//
// 규칙은 iOS(DropCore.NoteHierarchy, BRU-60)와 같다 — 같은 노트를 두 앱이 다르게
// 보여 주면 어느 쪽이 맞는지 알 수 없다. 다른 것은 들여쓰기 깊이 제한 하나뿐이다:
// iOS는 좁은 화면이라 2단에서 멈추고, 데스크톱은 넓어서 제한하지 않는다.

/** 계층 계산이 실제로 들여다보는 필드만 요구한다 — 테스트가 Note 전체를 만들 필요가 없게 */
export interface HierarchicalNote {
  id: string
  parentId: string | null
  createdAt: Date
}

export interface NoteRow<T extends HierarchicalNote = HierarchicalNote> {
  note: T
  /** 들여쓰기 단수. 0이면 최상위 */
  depth: number
  /** 필터에 걸린 것이 아니라 자식의 맥락을 위해 끌어온 노트 */
  isContextOnly: boolean
  /** 부모가 이 목록에 없어 최상위로 올라온 답글 */
  isOrphanedReply: boolean
}

/**
 * @param visible 필터·검색까지 통과해 실제로 보여야 하는 노트. 이 순서가 최상위 순서가 된다.
 * @param context 같은 뷰(활성/보관/휴지통)의 노트 전부. 부모를 끌어올 후보다.
 *
 * 1. 보이는 답글의 부모는 필터에 걸리지 않아도 끌어온다 — 그러지 않으면 답글이
 *    어느 쪽에도 못 들어가 목록에서 조용히 사라진다(이 이슈의 증상).
 * 2. 부모가 이 뷰에 아예 없으면 답글을 버리지 않고 최상위로 올린다.
 * 3. 형제 답글은 오래된 것부터 — 종전 동작 그대로다.
 */
export function buildNoteRows<T extends HierarchicalNote>(
  visible: T[],
  context: T[]
): Array<NoteRow<T>> {
  if (visible.length === 0) return []

  const byId = new Map<string, T>()
  const order: string[] = []
  for (const note of [...context, ...visible]) {
    if (byId.has(note.id)) continue
    byId.set(note.id, note)
    order.push(note.id)
  }

  const visibleIds = new Set(visible.map((note) => note.id))

  // 규칙 1 — 조상을 맥락으로 끌어온다
  const included = new Set(visibleIds)
  for (const id of visibleIds) {
    const seen = new Set([id])
    let current = byId.get(id)
    while (current?.parentId) {
      const parent = byId.get(current.parentId)
      // 망가진 데이터(순환)에서 멈추지 않으면 무한 루프가 된다
      if (!parent || seen.has(parent.id)) break
      seen.add(parent.id)
      included.add(parent.id)
      current = parent
    }
  }

  // 부모를 따라 올라가다 제자리로 돌아오는 노트를 가려낸다.
  // 자식으로 붙이면 어느 뿌리에서도 닿지 못해 목록에서 사라진다.
  const loops = new Set<string>()
  for (const id of included) {
    const seen = new Set([id])
    let current = byId.get(id)
    while (current?.parentId && included.has(current.parentId)) {
      const parent = byId.get(current.parentId)
      if (!parent) break
      if (seen.has(parent.id)) {
        loops.add(id)
        break
      }
      seen.add(parent.id)
      current = parent
    }
  }

  const childrenByParentId = new Map<string, T[]>()
  const rootIds: string[] = []
  for (const id of order) {
    if (!included.has(id)) continue
    const note = byId.get(id)
    if (!note) continue

    if (note.parentId && !loops.has(id) && included.has(note.parentId)) {
      const siblings = childrenByParentId.get(note.parentId) ?? []
      siblings.push(note)
      childrenByParentId.set(note.parentId, siblings)
    } else {
      rootIds.push(id)
    }
  }

  for (const [parentId, children] of childrenByParentId) {
    childrenByParentId.set(
      parentId,
      [...children].sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
    )
  }

  const emitted = new Set<string>()
  const rows: Array<NoteRow<T>> = []

  const emit = (note: T, depth: number): void => {
    if (emitted.has(note.id)) return
    emitted.add(note.id)

    rows.push({
      note,
      depth,
      isContextOnly: !visibleIds.has(note.id),
      // 부모가 있는데 최상위로 나왔다 = 부모를 잃은 답글
      isOrphanedReply: depth === 0 && note.parentId !== null && note.parentId !== note.id,
    })

    for (const child of childrenByParentId.get(note.id) ?? []) {
      emit(child, depth + 1)
    }
  }

  for (const id of rootIds) {
    const note = byId.get(id)
    if (note) emit(note, 0)
  }

  return rows
}
